daemon/listeners/listeners_linux.go

```diff
@@ -32,27 +36,34 @@ func Init(proto, addr, socketGroup string, tlsConfig *tls.Config) ([]net.Listene
 		}
 		ls = append(ls, l)
 	case "unix":
-		gid, err := lookupGID(socketGroup)
+		gid := os.Getgid()
+		l, err := sockets.NewUnixSocket(addr, gid)
 		if err != nil {
-			if socketGroup != "" {
+			return nil, errors.Wrapf(err, "can't create unix socket %s", addr)
+		}
+		if socketGroup != "" {
+			out, err := exec.Command("chgrp", socketGroup, addr).CombinedOutput()
```

Comment:
maybe we could do this only in the fallback if lookup with go implementation fails.

Reply:
@tonistiigi can you please elaborate why do you want to have both implementations? In case chgrp binary is not available, or as an optimization?

Reply:
I was asking to distinguish between the two cases because if you want an optimization (skip exec'ing chgrp if possible) I think in some cases `lookupGID()` might return some wrong result (say on a system with NIS or somesuch enabled), and in this case we'll end up with a wrong socket group, thereas `chgrp` should always be right.

If, on the other hand, you are concerned that `chgrp` binary might not be available, when we can amend the current code by falling back to `lookupGID`.

Reply:
...but I think a POSIX system lacking chgrp is something extremely weird, and an exit with an error seems like a proper thing to do in such case.

Reply:
I agree with @tonistiigi. Having the Go implementation as a default when Docker isn't statically linked does seem to be a good idea. Docker already relies on some binaries being present on the systems. It doesn't seem to be a good idea to introduce new requirements. The packages would have to be adjusted as well. POSIX systems may indeed have chgrp. Containers and other such stripped down environments may not.